### DISCUSSION TITLE
ZIP File Encryption

### DISCUSSION THREAD

**Discussion by Plecra** [zip-rs/zip-old#345] [Feature Design]
When creating a ZIP archive, we have a few options for encrypting the contents:

- traditional PKWARE encryption/"ZipCrypto"
  - Insecure
- Strong Encryption
  - Patented with no clear licensing option
- AE-x
  - fractured support from zip utilities (Windows Explorer can't read them)

For this project, AE-x is the only standard we could practically implement, however, even with the stronger algorithm, it doesn't obscure file metadata.  Since this could be taken advantage of, users who need their data to be secure are recommended to encrypt the whole archive.

Given all of these issues, I don't think there is a clear choice for an encryption API in the crate.

So for now, this issue is tracking any use cases for these features. If there is a real need for AE-x/ZipCrypto, we can further explore the design.

**Comment by BenjaminRi** [comment]
When it comes to crypto, I think we should aim for feature parity with 7-zip ( https://www.7-zip.org/ ). 7-zip does a lot of things right, the set of features it supports is among them (and because it's open-source, it's easy to get some inspiration if we get stuck). When I open the "Add to archive..." dialog, I get the following window:

![7zip-dialog](https://user-images.githubusercontent.com/28017860/86041068-62201180-ba45-11ea-85e9-08f1ad5d6f6a.png)

They support ZipCrypto and AES256.

- ZipCrypto is still surprisingly common. Despite the existence of various cryptographic attacks, it remains somewhat popular and widespread. We have ZipCrypto decryption now. Doing encryption isn't a big effort anymore. We need some cryptographically secure randomness, but I've looked into the `rand` crate (which we already depend on anyway) and it seems to provide such randomness. _However, should we even provide ZipCrypto encryption, knowing it is insecure?_

- AES256 is often supported and strong. Depending on how up to date your Linux distro is, the regular `unzip` tool can do it. Yes, the metadata is not encrypted, but that's perfectly fine if you e.g. AES256 encrypt your Bitcoin wallet private key. To support AES256, we have to introduce some new dependency that handles the algorithm such that it is secure and has good performance. AES256 is often directly supported by the CPU, many provide specific AES instructions. We don't want to implement this ourselves. It'll probably come down to an optional feature flag, possibly linking a C library. Though Rust crypto has come a long way, so maybe we can depend straight on a Rust library (I'd prefer that). We may also need stuff like SHA1 and SHA256 and so on, I haven't looked into that. It would be convenient to depend on a library that provides that for us too. Then we need to write the code to make it work with Zip (headers, password validation, etc.). That'll probably be a bit of non-trivial work. We have to be particularly careful with encryption because even minor mistakes can render the archive insecure. There are countless subtle pitfalls when implementing stuff like that. But by now, the specification is mature and that will help.

I think if we support those two algorithms, we support most crypto needs for Zip archives.

**Comment by mvdnes** [comment]
I would strongly discourage encrypting with ZipCrypto as it serves no good purpose due to its weaknesses. Decrypting is fine, to be compatible with existing archives.

**Comment by Plecra** [comment]
The [`aes-gcm`](https://docs.rs/aes-gcm/0.6.0/aes_gcm/) crate has been audited, and seems to be high quality. I'd be fine using that if we were to implement `AE-X` encryption. 

I'm still not sure it's worth it though - I don't think it's appropriate for a tool to claim to "encrypt your data" while leaving important information in plaintext. 

I think it would be better to encourage users to write to an encrypted `io::Write` (if they need security) and make sure our `ZipWriter` doesn't require `io::Seek`.  

Edit: to be clear, we should absolutely be able to *read* `AE-X` archives.

**Comment by Plecra** [comment]
I'd be receptive to a PR implementing a (non-default) `insecure-encryption` feature if it turns out that it has its use cases (there's bound to be an API out there that requires ZipCrypto). Due to the weaknesses of the spec, we also wouldn't demand a particularly defensive implementation.

**Comment by damccull** [comment]
Well I would love to be able to open zips with some or all of the files aes encrypted. Don't need to write it for my purpose but either seems impossible to do in rust.

**Comment by Plecra** [comment]
https://github.com/zip-rs/zip/pull/203 addresses decryption, and looks like a reasonable implementation so far. I'll need to fix the conflicts with the other PRs that have been merged in the meantime.

**Comment by mbr** [comment]
#203 has been cleaned up and rebased in the meantime, courtesy of @Lireer. Maybe it is time to give it another look? :)

**Comment by zamazan4ik** [comment]
I will look into the https://github.com/zip-rs/zip/pull/203

**Comment by KaDw** [comment]
Is there a way to handle password protected ZipCrypto encrypted file using this library? I have a legacy configuration file that is still using this type of encryption at some point and I'm thinking how I solve this using only rust

**Comment by Plecra** [comment]
Yup, the decrypting `ZipArchive` APIs can handle zipcrypto. You'll need to provide your password to the relevant function

**Comment by KaDw** [comment]
I have to both decrypt and encrypt password protected file. While decrypting works great I think encryption is missing

**Comment by Plecra** [comment]
Yes, so far zipcrypto encryption is deliberately missing from the API. Can you explain the application you need it for?

**Reply by KaDw** [reply]
Sure, there is an old game that store it's configuration files in password protected archive zipped using pkzipc. In addition a lookup table is used to further decode/encode the files and get the config in plaintext.

My goal is to create a simple tool to simply decode the files change the config and encode it again.

**Reply by BenjaminRi** [reply]
PKZIP decryption is there, encryption is not. All the parts are existing and I even wrote some stubs that just have to be finished (see: https://github.com/zip-rs/zip/blob/master/src/zipcrypto.rs#L43 ). It would be easy to finish, but the crate maintainers do not want it.

**Reply by Plecra** [reply]
I've added decryption as an unstable feature in 0.6.5 :) Would be great to hear if it does the trick for you!

```rs
// here's usage
use zip::unstable::write::FileOptionsExt;
let options = zip::write::FileOptions::default()
    .with_deprecated_encryption(b"password");
archive.start_file("name", options).unwrap();
archive.write_all(b"test").unwrap();
```

It's a bit awkward for now, and needs to buffer the entire file to allow the encryption to be correct. We could probably add Info-Zip's streaming decryption to fix that.

**Reply by KaDw** [reply]
Thats great! I will give it a try in a free time and let you know the results!

**Reply by damccull** [reply]
You're a very responsive dev.. Thanks for this amazing library!

**Reply by KaDw** [reply]
@Plecra It works! Thanks for this super quick fix, now I can do everything in pure Rust without any 3rd party tools like 7zip or pkzipc

**Reply by damccull** [reply]
I'm actually very excited to mess around with this too...
